atomicfu/src/concurrentMain/kotlin/kotlinx/atomicfu/locks/ParkingSupport.kt

```diff
@@ -0,0 +1,122 @@
+ *
+ * The caller is responsible for verifying the reason of wakeup and how to respond accordingly.
+ *
+ * Example usage parking thread:
+ * ```Kotlin
+ * // publish my parking handle
+ * handleReference.value = ParkingSupport.currentThreadHandle()
+ * // wait
+ * while (state.value == WAIT) {
+ *   ParkingSupport.park(Duration.INFINITE)
+ * }
+ * ```
+ *
+ * Example usage unparker thread:
+ * ```Kotlin
+ * state.value = WAKE
+ * ParkingSupport.unpark(handleReference.value)
+ * ```
+ *
+ * PLEASE NOTE: this is a low-level API and should be used with caution.
+ * Unless the goal is to create a _synchronization primitive_ like a mutex or semaphore,
+ * it is advised to use a higher level concurrency API like `kotlinx.coroutines`
+ */
+@ExperimentalThreadBlockingApi
+expect object ParkingSupport {
+
+    /**
+     * Parks the current thread for [timeout] duration.
+     *
+     * Wakes up in the following cases:
+     * - A different thread calls [ParkingSupport.unpark].
+     * - The [timeout] is exceeded.
+     * - A spurious wakeup
+     * - (Only on JVM) The thread was interrupted. The interrupted flag stays set after wakeup.
+     * A future call to [park] this thread will return immediately, unless the `Thread.interrupted` flag is cleared.
+     */
+    fun park(timeout: Duration)
+
+    /**
+     * Parks the current thread until [deadline] is reached.
+     *
+     * Wakes up in the following cases:
+     * - A different thread calls [ParkingSupport.unpark].
+     * - The given [deadline] has passed.
+     * - A spurious wakeup
+     * - (Only on JVM) The thread was interrupted. The interrupted flag stays set after wakeup.
+     * A future call to [park] this thread will return immediately, unless the `Thread.interrupted` flag is cleared.
+     */
+    fun parkUntil(deadline: TimeMark)
+
+    /**
+     * Unparks the thread corresponding to [handle].
+     * If [unpark] is called while the corresponding thread is not parked, the next [park] call will return immediately
+     * — the [ParkingHandle] is unparked ahead of time.
+     *
+     * A [ParkingHandle] can only _remember_ one pre-unpark attempt at a time.
+     * Meaning, when two consecutive [unpark] calls are made while the corresponding thread is not parked,
+     * only the next park call will return immediately — [unpark] calls are not accumulated.
+     */
+    fun unpark(handle: ParkingHandle)
+
+    /**
+     * Returns the [ParkingHandle] corresponding to the current thread.
+     * This [ParkingHandle] should be shared with other threads which allow them to [unpark] the current thread.
+     *
+     * A [ParkingHandle] is uniquely associated with a specific thread, maintaining a one-to-one correspondence.
+     * When the _same_ thread makes multiple calls to [currentThreadHandle],
+     * it always returns the _same_ [ParkingHandle].
+     *
+     * Note: as this function returns a unique [ParkingHandle] for each thread it should not be cached or memoized.
```

Comment:
But the example illustrating how to use parking caches (well, publishes, but its hard to tell the difference) the handle:
```
 * // publish my parking handle
 * handleReference.value = ParkingSupport.currentThreadHandle()
```

Reply:
I've changed it. Hope it is clearer this way. 

Otherwise, we could maybe remove this note all together since we already mentioned there is a one-to-one correspondance between a thread and the handle?

Reply:
It's much better now, thanks!

And I think it's better to leave the note as it might help some users.